Document IProvidePartitionKeyFromSagaId (#5148)
Particular/docs.particular.net: persistence/azure-table/configuration_code_astp_[3,).partial.md

snippet: EnableInstallersConfigurationOptingOutFromTableCreation

### Partitioning

> When storing saga's, saga IDs are by default deterministically derived from: the saga data, the correlation property name and the correlation property value. This deterministic ID is used as the partition key.

Reviewer: ```suggestion
When storing sagas, saga IDs are by default deterministically derived from the saga data:  the correlation property name and the correlation property value. This deterministic ID is used as the partition key.
```

> To deviate from that, the `IProvidePartitionKeyFromSagaId` interface can be injected into behaviors in the logical pipeline stage to set the partition key.

Reviewer: > To deviate from that

Not sure I follow what do you want to say here. Did you mean 

```suggestion
To set a custom partition key derived from saga ID, the `IProvidePartitionKeyFromSagaId` interface implementation can be injected into behaviors in the logical pipeline stage.
```

> snippet: BehaviorUsingIProvidePartitionKeyFromSagaId

Reviewer: The snippet seems to be missing, CI is failing.

snippet: EnableInstallersConfigurationOptingOutFromTableCreation

### Partitioning

When storing sagas, saga IDs are by default deterministically derived from the saga data:  the correlation property name and the correlation property value.
This deterministic ID is used as the partition key. The partition key is the key to achieving transactionality.

There might be scenario's in which it's desirable to set a custom partition key:

> - When upgrading to the Azure Table Persistence, it's possible make use of the added transactionality-features by setting the partition key for existing sagas that may have in-flight messages

Reviewer: ```suggestion
- When upgrading to the Azure Table Persistence, it's possible to make use of the added transactional features by setting the partition key for existing sagas that may have in-flight messages
```

> - To enable business operations to participate in transactions before the saga has been created, the saga ID can be precomputed and set as the partition key

Reviewer: It's a single transaction, right?

```suggestion
- To enable business operations to participate in the transaction before the saga has been created, the saga ID can be precomputed and set as the partition key
```

> ### Partitioning
> 
> When storing sagas, saga IDs are by default deterministically derived from the saga data:  the correlation property name and the correlation property value.
> This deterministic ID is used as the partition key. The partition key is the key to achieving transactionality.
> 
> There might be scenario's in which it's desirable to set a custom partition key:
> - When upgrading to the Azure Table Persistence, it's possible to make use of the added transactional features by setting the partition key for existing sagas that may have in-flight messages
> - To enable business operations to participate in the transaction before the saga has been created, the saga ID can be precomputed and set as the partition key
> 
> To set a custom partition key, the `IProvidePartitionKeyFromSagaId` interface implementation can be injected into behaviors in the logical pipeline stage.
> 
> snippet: BehaviorUsingIProvidePartitionKeyFromSagaId

Reviewer 2: ```suggestion

### Partitioning and compatibility helpers

In order to atomically store data within the same transaction all data written as part of the same message handling pipeline needs to share the same partition key. When storing sagas, saga IDs are by default deterministically derived from the saga data: the correlation property name and the correlation property value. So for conversations involving a saga as part of the message handling pipeline this determinstically derived ID might be a perfect fit as a partition key. Unfortunately the saga ID assignment happens quite late during the message handling and is not exposed to the user code. This makes it difficult to:

- Support transactional processing for existing sagas that where stored with the previous version of the persister and may have in-flight messages -or-
- Expose the deterministically created saga ID early enough to enable business operations to participate in the transaction before the saga has been created

To support the above scenarios the `IProvidePartitionKeyFromSagaId` interface implementation can be injected into behaviors in the logical pipeline stage.

snippet: BehaviorUsingIProvidePartitionKeyFromSagaId

The component will make sure to

- Derive the table to write the business data and the saga data to by looking for the necessary `TableInformation` or falling back to the default table
- Extract the saga id header if present and populate that header value as the partition key for previously created sagas
- When no header information is available and the compatibility mode is enabled do the necessary secondary index lookups and depending on the configuration even fallback to table scanning  and populate the found saga ID as the partition key. -or-
- When no header information is available and the compatibility mode is disabled precompute the deterministc saga ID based on the provided correlation propert information and populate the precomputed value as the partition key.
```

Reviewer 2: @lailabougria @SeanFeldman what do you think about this one? Might need to grammar tweaks though

Author: I incorporated your suggestion and made some grammar/wording changes @danielmarbach

snippet: EnableInstallersConfigurationOptingOutFromTableCreation

> ### Partitioning

Reviewer 2: I will try to reword this tomorrow. I think we need to make this clearer what the use case is. the intent was:

- enable partitioning and tx for customers that used previous version of the persistence and have sagas with messages in flight
- simple way to achieve partitioning and tx for conversations when sagas are involved to have the capability to precompute the saga id upfront so that business tx can participate even before the saga persister is at play

Author: Made some changes, please take a look
